### DISCUSSION TITLE
Deferred action on a closed socket

### DISCUSSION THREAD

**Discussion by arroyo-pl** [uNetworking/uWebSockets#1822] [Q&A] [answered]
Hi!

I have a scenario with `ws` as a pointer to a `WebSocket` object and two threads: a worker thread and a main thread (running the uWS event loop).

The issue occurs only under a fair load (100+ connections).
- **Worker thread**: Adds a deferred operation to **ws** using `loop->defer([ws] { ws->write("test"); });`.
- Meanwhile, in the **main thread** (event loop):
    - WebSocket is disconnected. I receive the `close` event but have no control over the deferred operation queue.
    - The loop attempts to perform the deferred operation on the disconnected `ws`, leading to a an AddressSanitizer error.
        
        
In such a situation, I would like a way to remove actions for the `ws` object, but how can I achieve that?

I have some ideas: I could create my own queue (instead of Loop's deferQueues), so I can remove actions for `ws` on close event, but I’m not sure how to execute it efficiently.
I tried hooking into `Loop::addPreHandler / addPostHandler`, but those are invoked at too long intervals. I also attempted to "wake up" the loop after adding something to the queue by using `us_wakeup_loop` to force the pre/postHandler execution, but it didn’t work.

I know I could use a timer via `us_create_timer`, but in this case, I need to set a fixed interval, e.g., 50ms. However, this would make the application less responsive. Setting an interval like 1ms would be excessive on the other hand.

Could you suggest some direction?
Thanks!

**Comment by uNetworkingAB** [comment]
Could possibly add ability to remove queued functions if the receiver holds
the queue mutex for the full duration of the execution.

But you can also just capture a shared_ptr to determine if the function is
still valid.

But yes, this is probably a good idea to add. If defer returns an integer
or something.

Need to think about it

Den tis 10 dec. 2024 20:34arroyo-pl ***@***.***> skrev:

**Comment by arroyo-pl** [comment]
Thanks!
Could you please say something more about shared_ptr? 
Currently, I only have a raw pointer to the WebSocket object, and (based on other discussions) there seems to be no other way to determine if the socket is valid except through the open/close events.

**Reply by uNetworkingAB** [reply]
In Boost Asio for instance, there is a design where all async operations are non-cancellable or at least non-reliably-cancellable. So what the design is there, is, you just capture a shared_ptr or a weak_ptr or something. And if that weak_ptr is no longer valid, you take it as cancelled. And so in your WebSocket UserData you have a shared_ptr or similar. So when it is destructed, the weak_ptr in flight is no longer valid

**Reply by uNetworkingAB** [reply, marked as answer]
Here's chatgpt reformulating it

You're describing a common design pattern in asynchronous programming with Boost.Asio, where the lifetime of operations and objects is managed using smart pointers. This approach works well to emulate cancellation by simply letting the owning shared pointer expire, which invalidates any weak pointers in flight. Here’s a breakdown of how and why this works:

The Design Rationale
Async Operations are Non-Cancellable: Boost.Asio doesn’t inherently support reliable cancellation of asynchronous operations. If you attempt to cancel an operation, it might still complete in some cases. The library’s design avoids guarantees for cancellation reliability to keep operations efficient and simple.

Shared Ownership: When an asynchronous operation is initiated, it often requires some user-defined context or state. This is encapsulated in a structure, commonly referred to as UserData, which is typically managed using std::shared_ptr.

Weak Pointer for Scoped Lifetime: When an operation is in progress, a std::weak_ptr to the UserData is held. If the original shared_ptr is destroyed (indicating the associated operation is no longer required), the weak_ptr becomes invalid (expired() returns true).

Using Weak Pointer for Cancellation: When the asynchronous operation checks the weak_ptr and finds it expired, it interprets this as a signal to abandon the operation or stop processing results. This provides an implicit cancellation mechanism without requiring explicit support from Boost.Asio.

**Reply by arroyo-pl** [reply]
Thanks, I'll try this approach.
